lib/resumption/resumable.go

```diff
@@ -0,0 +1,339 @@
+// Teleport
+// Copyright (C) 2023  Gravitational, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package resumption
+
+import (
+	"bufio"
+	"context"
+	"encoding/binary"
+	"io"
+	"net"
+	"sync/atomic"
+	"time"
+
+	"github.com/gravitational/trace"
+	"golang.org/x/sync/errgroup"
+)
+
+func newResumableConn(localAddr, remoteAddr net.Addr) *Conn {
+	r := &Conn{
+		managedConn: managedConn{
+			localAddr:  localAddr,
+			remoteAddr: remoteAddr,
+		},
+	}
+	r.cond.L = &r.mu
+	return r
+}
+
+// Conn is a [net.Conn] whose underlying transport can be closed and reopened,
+// to maintain the illusion of a perfect unbroken stream of bytes even if
+// network conditions would otherwise terminate a normal connection.
+type Conn struct {
+	managedConn
+
+	// requestDetach is non-nil if and only if there is an underlying connection
+	// attached; calling it should eventually result in the connection becoming
+	// detached, signaled by the field becoming nil.
+	requestDetach func()
+}
+
+var _ net.Conn = (*Conn)(nil)
+
+const handshakeTimeout = 5 * time.Second
+
+const (
+	errorTag        = ^uint64(0)
+	errorTagUvarint = "\xff\xff\xff\xff\xff\xff\xff\xff\xff\x01"
+)
+
+const maxFrameSize = 128 * 1024
+
+// runResumeV1Unlocking runs the symmetric resumption v1 protocol for r, using
+// nc as the underlying transport. The previous attached transport, if any, will
+// be detached immediately. firstConn signifies that the connection has not been
+// used, and the initial handshake will be assumed to be 0 for both sides. The
+// connection lock is assumed to be held when entering the function, since the
+// correct behavior of firstConn requires no possible external interference
+// before the attach point is reached; the lock will be not held when the
+// function returns.
+func runResumeV1Unlocking(r *Conn, nc net.Conn, firstConn bool) error {
+	defer nc.Close()
+
+	if !firstConn {
+		for !r.remoteClosed && r.requestDetach != nil {
+			r.requestDetach()
+			r.cond.Wait()
+		}
```

Comment:
is this dead code? Below we defer a function that sets `r.requestDetach = nil`

Reply:
Before returning we will set `r.requestDetach` to `nil`, but there might be more than one call of `runResumeV1Unlocking` running at the same time; this loop is specifically to ask the currently attached run to detach and leave the connection to us.

Remember that at this point we're holding the lock on `r`, but `r.cond.Wait()` will unlock it and let some other goroutine proceed on it until a `r.cond.Broadcast()`, and we, in fact, broadcast after setting `requestDetach` to `nil`.